File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.java
class Solution {
    public String rev(String s) {
        String str = "";
        for (int i = s.length() - 1; i >= 0; i--) {
            str += s.charAt(i) + "";
        }
        return str;
    }
    public int longestPalindromeSubseq(String s) {
        String str = rev(s);        
        int n = s.length();
        int[][] dp = new int[n+1][n+1];
        for(int i = 1; i<=n ; i++){
            for(int j = 1 ; j<=n ; j++){
                if(s.charAt(i-1)==str.charAt(j-1))  dp[i][j] =  1 + dp[i-1][j-1];
                else dp[i][j] =  Math.max(dp[i-1][j],dp[i][j-1]);
            }
        }
        return dp[n][n];
    }
}

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.cpp
class Solution {
public:
    int dp[1001][1001];
    int lcs(int i,int j,string& s,string& rev){
        if(i>=s.size()||j>=rev.size()) return 0;
        if(dp[i][j]!=-1) return dp[i][j];
        if(s[i] == rev[j]) return dp[i][j] = 1+lcs(i+1,j+1,s,rev);
        else return dp[i][j] = max(lcs(i+1,j,s,rev),lcs(i,j+1,s,rev));
    }
    int longestPalindromeSubseq(string s) {
        string rev = s;
        memset(dp,-1,sizeof(dp));
        reverse(rev.begin(),rev.end());
        return lcs(0,0,s,rev);
    }
};